test/extended/util/client.go

```diff
@@ -111,62 +109,63 @@ type resourceRef struct {
-		adminConfigPath:         KubeConfigPath(),
-		staticConfigManifestDir: StaticConfigManifestDir(),
+// CliOptions allows customization of the newely created CLI resource
+type CliOptions struct {
+	// KubeFramework to use, instead of creating a brand new one
+	KubeFramework *framework.Framework
+	// prefix for the created test namespace
+	BaseName string
+	// pod security level applied to the created e2e test namespace, defaults to `restricted`
+	PodSecurityEnforceLevel admissionapi.Level
+	// whether to add --namespace flag in oc invocations
+	WithoutNamespace bool
+	// whether we're dealing with HyperShift cluster
+	HyperShift bool
+}
+
+// NewCLI is responsible for creating a new CLI object for invoking OpenShift CLI,
+// CliOptions allow customization of the created object.
+func NewCLI(options CliOptions) *CLI {
+	if len(options.BaseName) == 0 && options.KubeFramework == nil {
+		panic("misconfigured CLI, either BaseName or KubeFramework are required")
 	}
-	// Called only once (assumed the objects will never get modified)
-	// TODO: run in every BeforeEach
-	cli.setupStaticConfigsFromManifests()
-	return cli
-}
-
-// NewCLIWithPodSecurityLevel initializes the CLI the same way as `NewCLI()`
-// but the given pod security level is applied to the created e2e test namespace.
-func NewCLIWithPodSecurityLevel(project string, level admissionapi.Level) *CLI {
-	cli := NewCLI(project)
-	cli.kubeFramework.NamespacePodSecurityEnforceLevel = level
-	return cli
-}
-
-// NewCLI initializes the CLI and Kube framework helpers with the provided
-// namespace. Should be called outside of a Ginkgo .It() function.
-// This will apply the `restricted` pod security level to the given underlying namespace.
-func NewCLI(project string) *CLI {
-	cli := NewCLIWithoutNamespace(project)
-	cli.withoutNamespace = false
-	// create our own project
-	g.BeforeEach(func() { cli.SetupProject() })
-	return cli
-}
-
-// NewCLIWithoutNamespace initializes the CLI and Kube framework helpers
-// without a namespace. Should be called outside of a Ginkgo .It()
-// function. Use SetupProject() to create a project for this namespace.
-func NewCLIWithoutNamespace(project string) *CLI {
-	cli := &CLI{
-		kubeFramework: &framework.Framework{
+	kubeFramework := options.KubeFramework
+	if kubeFramework == nil {
+		kubeFramework = &framework.Framework{
 			SkipNamespaceCreation: true,
-			BaseName:              project,
+			BaseName:              options.BaseName,
 			Options: framework.Options{
 				ClientQPS:   20,
 				ClientBurst: 50,
 			},
 			Timeouts: framework.NewTimeoutContext(),
-		},
+		}
+	}
+	if len(options.PodSecurityEnforceLevel) > 0 {
+		kubeFramework.NamespacePodSecurityEnforceLevel = options.PodSecurityEnforceLevel
+	}
+	var kubeConfig string
+	if options.HyperShift {
+		var err error
+		kubeConfig, _, err = GetHypershiftManagementClusterConfigAndNamespace()
+		o.Expect(err).NotTo(o.HaveOccurred())
+	} else {
+		kubeConfig = KubeConfigPath()
```

Comment:
CLIOptions should provide an indicator for
1. use the global
2. use this path
3. use the hypershift thing above

Reply:
Yeah, probably this comment along with the other about using `With*` methods should go together.